pkg/m3admin/commmon.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3admin
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+
+	retryhttp "github.com/hashicorp/go-retryablehttp"
+)
+
+// FormatHostname provides a formatted URL for HTTP reqeusts
+func FormatHostname(domain string) string {
+	return fmt.Sprintf("http://%s.%s:%d", M3coordinatorServiceName, domain, M3coordinatorPort)
+}
+
+// DoHttpReqeust is a simple helper for HTTP requests
+func DoHttpRequest(
+	client *retryhttp.Client,
+	action, url string,
+	data *bytes.Buffer,
+) (*http.Response, error) {
+	var request *retryhttp.Request
```

Comment:
unfamiliar with the retryhttp apis - is there a reason you have to break the code up into the two conditions as you have?

Would the version below work -- 
```
request, err := retryhttp.NewRequest(action, url, data)
if err != nil {
return nil, err 
}
request.Header...
...
}
```

Reply:
The check for `nil` is due to the slice `bytes` array can be `nil` however it is still typed as a `byte` slice and the condition ensure that untyped `nil` is passed to the receiver.

Reply:
hm could you avoid this issue if you just expected `data []byte` instead of `data *bytes.Buffer`?

Reply:
```
// NewRequest creates a new wrapped request.                                   
func NewRequest(method, url string, rawBody interface{}) (*Request, error) {   
  var err error                                                                
  var body ReaderFunc                                                          
  var contentLength int64                                                         
                                                                                  
  if rawBody != nil {                                                             
    switch rawBody.(type) {  
....
```

Not sure, it's an untyped `interface` that is evaluated within a `switch` for several variants in which `[]byte` is one of the types that is cased.